### ISSUE TITLE
METABUG: Tier 2 optimizer

### ISSUE THREAD

**Issue by mdboom** [faster-cpython/ideas#587]
> **Note**
> This issue is not for discussion, but to keep the various steps organized. To create new subtasks, create new issues and link to them from here by editing this comment.

To discuss the Tier 2 optimizer topic in general, see #557.

### Components

* VM hooks and callbacks -- Enable VM to tell optimizer about hotspots and to warn it of changes.
* #558
* #560
* #561
* #580
* #563

### Design artifacts

* Optimizer API
* Optimizer instruction format
* Tier 2 interpreter instruction format (if different from the optimizer instruction format)
* Executable object layout

### Dependencies

```mermaid
graph TD;
  optimizer_api[Optimizer API] --> superblock_generator[Superblock generator];
  optimizer_format[Optimizer instruction format] --> superblock_generator;
  superblock_generator --> specializer[Specializer];
  superblock_generator --> partial_evaluator[Partial evaluator];
  superblock_generator --> compiler[Compiler];
  superblock_generator --> deoptimization[Deoptimization of superblocks];
  deoptimization[Deoptimization of superblocks] --> partial_evaluator[Partial evaluator];
  optimizer_format --> interpreter_format[Interpreter instruction format];
  interpreter_format --> interpreter[Interpreter];
  executable_object_layout[Executable object layout] --> interpreter;
  executable_object_layout[Executable object layout] --> compiler;
  interpreter --> specializer;
  interpreter --> partial_evaluator;
  interpreter --> compiler;
```

**Comment by markshannon** [comment]
Although the development of the partial evaluator and compiler can proceed in parallel with that of the specializer, it is noting that the partial evaluator and compiler have little value without the specializer. The specializer must take priority over the other optimizer components.

**Comment by markshannon** [comment]
With the exception of the partial evaluator, which has its own issue, the component parts of this are done or obsolete
